### ISSUE TITLE
MultiShardSample uses same seed on every worker

### ISSUE THREAD

**Issue by iceboundflame** [webdataset/webdataset#242]
`MultiShardSample` looks quite handy for combining multiple datasets with different weights, sampling shards from some datasets more frequently than others.

I found that its current implementation appears to use the same RNG for every worker, and it doesn't seed according to the worker ID. That means when using PyTorch DataLoader workers, each worker process unfortunately gets the same set of shards.

The README.md only hints at this capability under `Loadable Dataset Specifications` so maybe it is still a WIP feature. I am doing the following:

```
spec = dict(datasets=[
            dict(name='a',
                 shards="pipe:s3cmd get s3://.../a-{001..006}.tar -",
                 resample=1),
            dict(name='b',
                 shards="pipe:s3cmd get s3://.../b-{001..009}.tar -",
                 resample=3),
])
ds = WebDataset(spec)
...
```

I am using webdataset version 0.2.31.

Thank you!

**Comment by SCZwangxiao** [comment]
I don't think the problem you mentioned is a bug, but a feature. 

You mentioned: `when using PyTorch DataLoader workers, each worker process unfortunately gets the same set of shards`. However, given that each worker has exactly the same shard list, we can easily split the shard list to every worker in every node, by simply slicing the shard list:

```python
import webdataset as wds

pipeline = [wds.MultiShardSample(self.input_shards)]
pipeline.extend([
    wds.split_by_node,
    wds.split_by_worker
])
```

where the split function is quite simple:
```python
def split_by_node(src, group=None):
    rank, world_size, worker, num_workers = utils.pytorch_worker_info(group=group)
    if world_size > 1:
        for s in islice(src, rank, None, world_size):
            yield s
    else:
        for s in src:
            yield s

def split_by_worker(src):
    rank, world_size, worker, num_workers = utils.pytorch_worker_info()
    if num_workers > 1:
        for s in islice(src, worker, None, num_workers):
            yield s
    else:
        for s in src:
            yield s

```

**Comment by titamar** [comment]
Hi, I encountered the same issue using WebDataset. When seed==None (default), it is currenlty converted to a random seed which is shared accross workeres. If I choose not to split the shards between workers, all workers share the same seed for the shard shuffle buffer. This makes all of them exactly the same and produce the same batches. 

I find this default behavior a bit unexpected. I would expect the seed to stay None if this is what I passed to the constructor. I think the user can pass a seed if he wants the deterministic behavior. By default, I suggest to just use the seed provided by the user, including seed=None (uses process id).
